src/mol-util/fibonacci-heap.ts

```diff
@@ -0,0 +1,388 @@
+/**
+ * Copyright (c) 2018-2022 mol* contributors, licensed under MIT, See LICENSE file for more info.
+ *
+ * @author Gianluca Tomasello <giagitom@gmail.com>
+ *
+ * Adapted from https://github.com/gwtw/ts-fibonacci-heap, Copyright (c) 2014 Daniel Imms, MIT
+ */
+
+type CompareFunction<K, V> = (a: INode<K, V>, b: INode<K, V>) => number;
+
+interface INode<K, V> {
+    key: K;
+    value?: V;
+}
+
+class Node<K, V> implements INode<K, V> {
+    public key: K;
+    public value: V | undefined;
+    public prev: Node<K, V>;
+    public next: Node<K, V>;
+    public parent: Node<K, V> | null = null;
+    public child: Node<K, V> | null = null;
+
+    public degree: number = 0;
+    public isMarked: boolean = false;
+
+    constructor(key: K, value?: V) {
+        this.key = key;
+        this.value = value;
+        this.prev = this;
+        this.next = this;
+    }
+}
+
+class NodeListIterator<K, V> {
```

Comment:
Can you add a `reset(node)` method here and use a static instance of this where possible?

Reply:
>See if the static iterator with reset helps with the fib heap?

I have made some tests with and without static iterator and seems that there is no difference when making a search with k=100.000 (about **300ms**). 
Instead there is a worsening in performances when making 100.000 different random coordinate nearest searches with k=1 (**18.8 seconds** with static iterator vs **16.7 seconds** without). 

Do you think this is something strange? 
I implemented static iterator reset method i this way 
```ts
public reset(start: Node<K, V>) {
        this._index = -1;
        this._items = []; // using this._items.length = 0 leads to about 1 second performances worsening 
        let current = start;
        do {
            this._items.push(current);
            current = current.next;
        } while (start !== current);
    }
```
And initilaized it like this.
```ts
const tmpIt = new NodeListIterator<any, any>();
```

Reply:
That seems not such a big improvement. The different would be likely that in your implementation with k=1, the iterator just gets created more often. But ~18s for k=1 100k queries seems to be quite slow. I reckon you can optimize it to be at least 10x faster.